File: wolf_nlp/alg_practice/hmm/hmm.py
#coding=utf8

import numpy as np

"""
    隐马模型是关于时序的概率模型，描述一个隐藏的马尔科夫链随机生成不可观测的状态随机序列，在生成一个观测序列的过程
    隐藏的马尔科夫随机生成的状态序列
"""

class HMM(object):
    def __init__(self,N,M):
        #状态转移
        self.A_=np.zeros((N,N))
        #发射概率
        self.B_=np.zeros((N,M))
        #初始状态概率矩阵
        self.pi_=np.array([1.0/N]*N)
        #状态集
        self.States_=N
        #观测集
        self.obs_=M
    def cal_prob(self, obs):
        res = self.forward(obs=obs)
        return sum(res[len(obs)-1])

    def forward(self, obs):
        self.alpha = np.zeros((len(obs), self.States_))

        #公式 10.15
        for i in range(self.N):
            #初始概率*发射概率
            self.alpha[0][i] = self.pi_[i] * self.B_[i][self.obs_[0]]

        # 公式 10.16
        for t in range(1,len(obs)):
            for i in range(self.N):
                sum_pro = 0.0
                for j in range(self.N):
                    sum_pro += self.alpha[t-1][j] * self.A_[j][i]
                self.alpha[t][i] = sum_pro * self.B_[i][obs[t]]

    def backward(self, obs):
        obs_len = len(obs)
        self.beta = np.zeros((obs_len, self.States_))

        #公式10.19
        for i in range(self.N):
            self.beta[obs_len-1][i] = 1

        #公式10.20
        for t in range(obs_len-2,-1,-1):
            for i in range(self.N):
                for j in range(self.N):
                    self.beta[t][i] += self.A_[i][j] * self.B_[j][obs[t+1]] * self.beta[t+1][j]

    def cal_gamma(self, t, i):
        #公式10.24 得到单个状态的概率计算公式，计算在时刻t处于状态q的概率
        numberator = self.alpha[t][i] * self.beta[t][j]
        denominator = 0.0
        for j in range(self.States_):
            denominator += self.alpha[t][i] * self.beta[t][j]
        return numberator/denominator

    def cal_ksi(self,t,i,j):
        #公式10.26  得到两个状态概率的计算公式，计算时刻t处于状态qi且时刻t+1处于状态qj的概率
        numberator = self.alpha[t][i] * self.A_[i][j] * self.B_[j][self.obs_[t+1]] * self.beta[t+1][j]
        denominator = 0.0
        for i in range(self.States_):
            for j in range(self.States_):
                denominator += self.alpha[t][i] * self.A_[i][j] * self.B_[j][self.States_[t+1]] * self.beta[t+1][j]
        return numberator/denominator


    def init(self):
        """
            随机生成A,B,pi
            保证每行相加等于1
        """
        import random
        for i in range(self.States_):
            randomlist = [random.randint(1,100) for _ in range(self.States_)]
            sum_tot = sum(randomlist)
            for j in range(self.States_):
                self.A_[i][j] = randomlist[j]/sum_tot

        for i in range(self.States_):
            randomlist=[random.randint(1,100) for _ in range(self.States_)]
            sum_tot = sum(randomlist)
            for j in range(self.obs_):
                self.B_[i][j] = randomlist[j]/sum_tot

    def train(self, obs, max_step=100):
        """
            鲍姆韦尔奇
        """
        obs_len = len(obs)
        self.init()
        step = 0
        while step < max_step:
            step += 1
            tmp_A = np.zeros((self.States_, self.States_))
            tmp_B = np.zeros((self.States_, self.obs_))
            tmp_pi = np.array([0.0] * self.States_)

            self.forward()
            self.backward()

            # a_{ij}
            for i in self.States_:
                for j in self.States_:
                    numberator = 0
                    denominator = 0
                    for t in range(obs_len):
                        numberator += self.cal_ksi(t,i,j)
                        denominator += self.cal_gamma(t, i)
                    tmp_A[i][j] = numberator/denominator

            # b{jk}
            for j in self.States_:
                for k in self.obs_:
                    numberator = 0
                    denominator = 0
                    for t in range(obs_len):
                        if k == obs[t]:
                            numberator += self.cal_gamma(t,j)
                        denominator += self.cal_gamma(t,k)
                    tmp_B = numberator/denominator

            #pi_i
            for i in range(self.States_):
                tmp_pi = self.cal_gamma(0,i)

            self.A_ = tmp_A
            self.B_ = tmp_B
            self.pi_ = tmp_pi

    def viterbi(self, obs):
        # 求出现obs最可能的隐状态序列（如词性标注）
        V = [{}]
        #初始化V
        for i in self.States_:
            V[0][i] = {'prob':self.pi_[i] * self.B_[i][obs[0]],'prev':None}

        max_prob = 0
        max_prob_state_pre = 0
        max_prob_state = 0
        for t in range(1,len(obs)):
            for i in range(self.States_):
                for j in range(self.States_):
                    p = V[t-1][j] * self.A_[j][i] * self.B_[i][obs[t-1]]
                    if max_prob > p:
                        max_prob = p
                        max_prob_state_pre = j
                        max_prob_state = i
                V[t][i] = {'prob':max_prob, "prev":max_prob_state_pre}

        obs_res = [max_prob_state]
        for t in range(len(V)-1, -1, -1):
            obs_res.append(0, V[t][max_prob_state_pre]["prev"])
            max_prob_state_pre = V[t][max_prob_state_pre]["prev"]


# 求最大的隐状态序列
def Viterbi(obs,states,start_prob,emit_prob,trans_prob):
    #先初始化变量
    max_states_value = max(states)
    max_states_value += 1
    # V[obs][states] = 概率  每一列存储第一列不同隐状态下的概率
    V = [0 for _ in range(max_states_value)] * len(obs)
    # path[states][obs] 当前状态是哪个隐状态 每一行存储V对应的路径
    path = [0 for _ in range(len(obs))] * max_states_value

    #初始化状态
    for i in states:
        V[0][i] = start_prob[i] + emit_prob[i][obs[0]]
        path[i][0] = i

    for t in range(1, len(obs)):
        new_path = [0 for _ in range(len(obs))] * max_states_value
        for y in states:
            prob = 0.0
            max_prob_state = 0
            for y0 in states:
                nprob = V[t-1][y0] + trans_prob[y0][y] + emit_prob[y][obs[t]]
                if nprob > prob:
                    prob = nprob
                    max_prob_state = y0
                    V[t][y] = prob
                    new_path[y][t] = y
        path = new_path

    prob = 0.0
    max_prob_state = 0
    for i in states:
        if V[len(obs)-1][y] < prob:
            prob = V[len(obs)-1][y]
            max_prob_state = y
    return path[max_prob_state]

def Viterbi2(obs, states, start_prob, trans_prob, emit_prob):
    V= [{}]
    # 先初始化V
    for i in states:
        V[0][i] = {"prob":start_prob[i] * emit_prob[i][obs[0]], "prev":None}

    #run Viterbi when t > 0
    for t in range(1, len(obs)):
        for y in states:
            max_prob = 0.0
            pre_state = 0
            for y0 in states:
                prob = V[t-1][y0]["prob"]*trans_prob[y0][y]*emit_prob[y][obs[t]]
                if prob > max_prob:
                    max_prob = prob
                    pre_state = y0
            V[t][y] = {"prob":max_prob, "prev":pre_state}
    opt = []
    pre = None
    max_prob = 0.0
    for state,data in V[-1].items():
        if data["prob"] > max_prob:
            max_prob = data["prob"]
            opt = [state]
            pre = state

    for t in range(len(V)-1, -1, -1):
        opt.insert(0, V[t][pre]["prev"])
        pre = V[t][pre]["prev"]

    print("max prob {0} path {1}".format(max_prob, " ".join(opt)))

def viterbi(obs, states, start_prob, trans_prob, emit_prob):
    V = [{}]
    #先初始化
    for i in states:
        V[0][i] = {'prob':start_prob[i] * emit_prob[i][obs[0]], 'prev':None}
    #开始进行viterbi计算
    for t in range(1, len(obs)):
        for y in states:
            max_prob = 0.0
            pre_state = None
            for y0 in states:
                #前一时刻不同状态的概率 * 不同状态转移到当前状态概率 * 当前状态的发射概率
                prob = V[t-1][y0]['prob'] * trans_prob[y0][y] * emit_prob[y][obs[t]]
                if prob > max_prob:
                    max_prob = prob
                    pre_state = y0
            V[t][y] = {'prob':max_prob, 'prev':pre_state}
    opt = []
    max_prob = 0.0
    pre_state = None
    #先找出最后一个最大概率的
    for p,s in V[-1].item():
        if p > max_prob:
            max_prob = p
            opt = [s]
            pre_state = s
    #从最后一个逆向遍历
    for t in range(len(V[-1])-1,-1,-1):
        opt.insert(0, V[-1][pre_state]['prev'])
        pre_state = V[-1][pre_state]['prev']

if __name__ == "__main__":
    Viterbi()

File: wolf_nlp/alg_practice/hmm/hmm.cpp
#include <iostream>
#include <stdint.h>
#include <vector>
#include <map>
#include <algorithm>
using namespace std;

/*
    参考：https://en.wikipedia.org/wiki/Viterbi_algorithm#Example
        https://www.jianshu.com/nb/18954627
*/

/*
    //http://www.52nlp.cn/hmm-learn-best-practices-five-forward-algorithm-4
    hmm一个模型，俩个假设，五个基本元素，4大问题，几种算法（前向、后向、EM、BaumWelch、viterbi）
    三大假设：
        1、齐次性假设（隐状态仅与前一个有关）
        2、观测独立性假设（输出只与当前状态有关）
*/

typedef struct {
    int32_t N;  //隐状态数目
    int32_t M;  //观测状态数目
    double **A;   //转移矩阵A[N][N]
    double **B;   //发射矩阵B[N][M]
    double *pi;    //初始概率分布pi[N]
}HMMInfo;

/*
    phmm：已知的HMM模型
    T：观测序列时间长度T
    O：观测序列
    alpha：局部概率，中间状态
    pprob：最终的观测概率
*/
void Forward(HMMInfo *phmm, int32_t T, int32_t *O, double **alpha, double *pprob)
{
    int32_t i,j; //状态索引
    int32_t t;  //时间索引
    double psum; //局部概率的中间值
    /*1、初始化：计算t=1时刻所有状态的局部概率alpha*/
    for(i = 1; i <= phmm->N; i++){
        alpha[1][i] = phmm->pi[i] * phmm->B[i][O[1]];
    }

    /*2、归纳：递归计算每个时间点的局部概率*/
    for(t = 1; t < T; t++){
        for(j = 1; j <= phmm->N; j++){
            psum = 0;
            for(i = 1; i <= phmm->N; i++){
                psum += alpha[t][i] * phmm->A[i][j];
            }
            alpha[t+1][j] = psum * phmm->B[j][O[t+1]];
        }
    }
    /*3、观测序列的概率等于T时刻所有局部概率之和*/
    *pprob = 0.0;
    for(i = 1; i < phmm->N; i++){
        *pprob += alpha[T][i];
    }
}

void Backward(HMMInfo *phmm, int32_t T, int32_t *O, double **beta, double *pprob)
{
    int32_t i,j;
    int32_t t;
    double psum;
    /*1、初始化*/
    for(i = 1; i <= phmm->N; i++){
        beta[T][i] = 1.0;
    }
    /*2、归纳*/
    for(t = T - 1; t >= 1; t--){
        for(i = 1; i <= phmm->N; i++){
            psum = 0.0;
            for(j = 1; j <= phmm->N; j++){
                psum += phmm->A[i][j] * phmm->B[j][O[t+1]] * beta[t+1][j];
            }
            beta[t][i] = psum;
        }
    }

    /*3、结束*/
    *pprob = 0.0;
    for(i = 0; i < phmm->N; i++){
        *pprob += beta[1][i];
    }
}

void BaumWelch(HMMInfo *phmm, int32_t t, int32_t *O, double **alpha, double **beta, double **gamma, int32_t *pniter, double *plogprobinit, double *plogprobfinal)
{
    /*
    ForwardWithScale();
    BackwardWithScale();
    ComputeGamma();
    ComputeXi();
    */
}

void Viterbi(HMMInfo *phmm, int32_t T, int32_t *O, double **delta, int32_t **psi, int32_t *q, double *pprob)
{
    int32_t i,j;
    int32_t t;
    int32_t maxvalind;
    double maxval,val;

    //1、初始化
    for(i = 1; i <= phmm->N; i++){
        delta[1][i] = phmm->pi[i] * phmm->B[i][O[1]];
        psi[1][i] = 0;
    }

    //2、归纳
    for(t = 2; t <= T; t++) {
        for(j = 1; j <= phmm->N; j++) {
            maxval = 0.0;
            maxvalind = 1;
            for(i = 1; i < phmm->N; i++) {
                val = delta[t-1][i] * phmm->A[i][j];
                if(val > maxval){
                    maxval = val;
                    maxvalind = i;
                }
            }
            delta[t][j] = maxval * phmm->B[j][O[t]];
            psi[t][j] = maxvalind;
        }
    }

    //3、结束
    *pprob = 0.0;
    q[T] = 1;
    for(i = 1; i<= phmm->N; i++){
        if(delta[T][i] > *pprob){
            *pprob = delta[T][i];
            q[T] = i;
        }
    }

    //4、回溯path
    for(t = T - 1; t > 0; i--){
        q[t] = psi[t+1][q[t+1]];
    }
}

class HMM
{
public:
    HMM(){}
    ~HMM(){}

public:
    void Init()
    {
        //初始化
    }

    //通过观测序列得到最可能的隐状态序列
    vector<int32_t> Viterbi(vector<int32_t> obs)
    {
        int32_t max_state_value = 0;
        for (auto &it : states_){
            max_state_value = max(max_state_value, it);
        }

        max_state_value++;
        //[时刻][隐状态]=概率
        vector<float> maxv_vec(max_state_value);
        vector<vector<float>> v;
        for (auto &it:obs) {
            v.push_back(maxv_vec);
        }

        //中间变量 每一行存储v对于路径
        vector<int32_t> vec_obs(obs.size());
        vector<vector<int32_t>> path;
        for (auto i = 0; i < max_state_value;i++) {
            path.push_back(vec_obs);
        }

        for (auto &it:states_) {
            v[0][it] = start_prob_[it] + emit_prob_[it][obs[0]];
            path[it][0] = it;
        }

        for (auto &t:obs) {
            //遍历当前的各种状态
            for (auto &curr : states_){
                //计算从上一刻转移到现在 最大概率，计算最大概率
                for (auto &last : states_){
                    //概率 = 上一时刻是last概率 * last转移到curr概率 * curr转移到观测的概率
                    float nprob = v[t - 1][last] + trans_prob_[last][curr] + emit_prob_[curr][obs[t]];
                }
            }
        }

        return vec_obs;
    }

    struct ViterbiInfo
    {
        int32_t t_;
        int32_t s_;
        float p_;
        int32_t pre_s_;
        ViterbiInfo() :t_(0), s_(0), p_(0.0), pre_s_(0){}
    };

    float Viterbi(vector<int32_t> obs, vector<int32_t> &des_path)
    {
        vector<map<int32_t, ViterbiInfo>> V;
        V.resize(obs.size());
        //初始化
        for (auto &it : states_){
            ViterbiInfo info;
            info.t_ = 0;
            info.s_ = it;
            info.p_ = start_prob_[it] * emit_prob_[it][obs[0]];
            info.pre_s_ = -1;
            V[0].insert(make_pair(it, info));
        }

        //开始遍历计算
        for (auto t = 1; t < obs.size(); t++) {
            for (auto &y : states_) {
                float max_prob = 0.0;
                int32_t max_prob_state_pre = 0;
                for (auto &y0 : states_) {
                    float p = V[t - 1][y0].p_ * trans_prob_[y0][y] * emit_prob_[y][obs[t]];
                    if (p > max_prob) {
                        max_prob_state_pre = y0;
                        max_prob = p;
                    }
                }

                ViterbiInfo info;
                info.t_ = t;
                info.s_ = y;
                info.p_ = max_prob;
                info.pre_s_ = max_prob_state_pre;
                V[t].insert(make_pair(y, info));
            }
        }

        //定位到最大概率的状态
        float max_prob = 0.0;
        int32_t max_prob_state_pre = 0;
        for (auto &it : V[V.size() - 1]) {
            if (it.second.p_ > max_prob){
                max_prob = it.second.p_;
                max_prob_state_pre = it.second.pre_s_;
            }
        }

        //遍历V找到最大概率对于的路径
    }

private:
    //状态集
    vector<int32_t> states_;
    //初始化概率
    vector<float> start_prob_;
    //转移概率A
    vector<vector<float>> trans_prob_;
    //发射概率B
    vector<vector<float>> emit_prob_;
};

//快排和归并
void quickSort(int32_t *source, int32_t l, int32_t r)
{
    if (l >= r){
        return;
    }

    int32_t key = source[l];
    int32_t begin = l;
    int32_t end = r;
    while (l < r) {
        while (l < r) {
            if (source[r] >= key) {
                r--;
            }
        }
        while (l < r){
            if (source[l] <= key){
                l++;
            }
        }

        swap(source[l], source[r]);
    }
    swap(source[begin], source[l]);
    quickSort(source, begin, l-1);
    quickSort(source, l+1, end);
}

int main()
{
    /*
    HMM hmm = HMM();
    hmm.Init();

    vector<int32_t> obsdata;
    obsdata.push_back(1);
    obsdata.push_back(2);
    obsdata.push_back(3);
    vector<int32_t> res = hmm.Viterbi(obsdata);
    for (auto it:res) {
        cout << it << " ";
    }
    cout << endl;
    */
    int32_t s[10] = {10,1000,30,500,300,600,200,100,400,80};
    cout << sizeof(s) << endl;
    //quickSort(s, 0, sizeof(s)-1);
    for (auto i = 0; i < 10; i++) {
        cout << s[i] << " ";
    }
    cin.get();
    return 0;
}
